src/OrchardCore.Modules/Orchard.Autoroute/Handlers/AutoroutePartHandler.cs

```diff
@@ -119,5 +124,29 @@ private string GetPattern(AutoroutePart part)
 
             return pattern;
         }
+
+        private string GenerateUniquePath(string path, AutoroutePart context)
+        {
+            var similarPaths = _session.QueryIndexAsync<AutoroutePartIndex>(o => o.ContentItemId != context.ContentItem.ContentItemId && o.Path.StartsWith(path)).List().GetAwaiter().GetResult();
```

Comment:
`GenerateUniquePath` is called only by `AutoroutePartHandler.Updated` method which isn't async, or am I missing something?

I am using `StartWith` because this approach requires only one DB query under all circumstances. 
* if the exact match isn't found, the generated path is unique and can be used
* if there is the exact match, I try to extract a version from returned paths, select the highest one and use  version + 1 for the generated path
* there is a risk, that query might return many records from DB if user tries to generate `Path` for very simple `Title` such `A`

I also considered a solution similar to yours, but it requires at least 2 DB queries and there is a risk, that it will require even more (one for every version, that is already used).

I like the first approach little more, but if you think the second one is better I will rewrite it.